### ISSUE TITLE
Persist watched mutations

### ISSUE THREAD

**Issue by miguelocarvajal** [haytko/apollo-link-watched-mutation#15]
Am playing with the idea of persisting the watched mutations due to some offline tests we are doing.

I think it would be reasonable in the following scenario:

1. The user opens a native app, the server is successfully queried (online)
2. The user closes the app
3. Device is brought offline and the user re-opens the app. (with the current implementation, the query will fail but using cache-and-network will allow user to see data)
4. Device is brought online without closing the app
5. Mutation is done and the callback will not be invoked

Currently the watched mutations are kept in memory and I understand this is done to keep track of the successful queries and their variables so that the callback function can be invoked for every successful query along with the corresponding variables.

I think it would make sense to implement the query manager in such a way that previously successful queries are persisted so that the callback is invoked in step 5 above.

What are your thoughts? I would prepare a PR if this makes sense for the project.

**Comment by haytko** [comment]
Hmm, sounds interesting, i'm pretty unfamiliar with offline use cases.

I guess my main question would be: 
- Are you using a cache such that when the device is brought back online in step 4 those queries from step 1 are still present ? If they are this makes sense to me since the app would continue to work as if it never closed or went offline, otherwise it doesn't make much sense to me.

I'm interested in seeing what a PR would look like but this would definitely need to be an optional feature. Also, I'd like to be aware of any security concerns as well since we're persisting potentially sensitive data.

👍 Thanks for contributing on these issues and ideas recently. I appreciate it.

**Comment by miguelocarvajal** [comment]
Yes, exactly. We do have the scenario where the device is brought back online and the original query data is still present (using apollo-cache-persist). We use this for apps where the user stays logged in and data is available until they manually log out.

What I am working on lets you create a custom query manager so that what is currently kept in-memory can be persisted via whatever means (window.localStorage, AsyncStorage, etc). This part of it is already done, there is one scenario that I need to figure out which involves issue #12. So we would need to sort that out before fully diving into this.

I do have the concern about security as well though, it would be ideal if we could delete this data when the Apollo cache is cleared out (clearStore()). I will have to look into this as well.

In the meantime I'll start cleaning up my code and prepare a PR so you can take a look.

**Comment by haytko** [comment]
I'd be really interested in seeing if we can build up the list of current queries when we construct the link. 

Can we just do a read from the cache then ? 

I think this would be my ideal solution because it means we don't need to worry about persisting anything, the cache is the ultimate source of truth.

**Comment by miguelocarvajal** [comment]
I thought about this, but in order to get the cache for the queries we need the variables used to execute the query.

I think it would be ideal to read the cache for the "active" queries, that is queries that actually have data in the cache.

Do you have any ideas of how we can do that?

**Comment by haytko** [comment]
Hmm, it might be a big ugly but I think and hope we can ? 

If we can't it gives me some hesitation with going this path because I think if I were on the apollo team the idea of links each persisting their own distributed copies of data sounds really untenable.

**Comment by miguelocarvajal** [comment]
I do agree I dislike the idea of copying data in a link, although what we're copying is not the what's in the cache. It's more of the "active queries".

I've looked into this and last that I saw the only way to get data in/out of the cache is via the readQuery/writeQuery methods. Both of these methods require the use of query document and variables, so how else would we get the information so we don't have to persist the watched mutations?

If we find a way to not have to distribute the data it would be great, although this is what apollo-link-watched-mutations is currently doing. I am just proposing that it gets persisted so that things work when the app is closed and opened again.
